src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModelFactory.cs

```diff
@@ -66,18 +63,19 @@ internal sealed class ProjectCodeModelFactory : IProjectCodeModelFactory
             _visualStudioWorkspace.WorkspaceChanged += OnWorkspaceChanged;
         }
 
-        private System.Threading.Tasks.Task ProcessNextDocumentBatchAsync(
+        private Task ProcessNextDocumentBatchAsync(
             ImmutableArray<DocumentId> documentIds, CancellationToken cancellationToken)
         {
             foreach (var documentId in documentIds)
             {
                 // Now, enqueue foreground work to actually process these documents in a serialized and incremental
                 // fashion.  FireEventsForDocument will actually limit how much time it spends firing events so that it
                 // doesn't saturate  the UI thread.
-                _notificationService.RegisterNotification(
-                    () => FireEventsForDocument(documentId),
-                    _listener.BeginAsyncOperation("CodeModelEvent"),
-                    cancellationToken);
+                Task.Run(async () =>
+                {
+                    await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+                    FireEventsForDocument(documentId);
+                }, cancellationToken).CompletesAsyncOperation(_listener.BeginAsyncOperation("CodeModelEvent"));
```

Comment:
@jasonmalinowski @sharwell is this the right way to translate to use JTF?

Reply:
This needs the same change I mentioned earlier